processor/src/operations/sys_ops.rs

```diff
@@ -122,9 +131,62 @@ impl Process {
     {
         self.stack.copy_state(0);
         self.decoder.set_user_op_helpers(Operation::Emit(event_id), &[event_id.into()]);
-        host.on_event(self.into(), event_id)?;
 
-        Ok(())
+        // If it's a native event, handle it directly. Otherwise, forward it to the host.
+        if let Some(advice_injector) = AdviceInjector::from_event_id(event_id) {
+            self.handle_advice_injector(advice_injector, host)
+        } else {
+            host.on_event(self.into(), event_id)
+        }
+    }
+
+    fn handle_advice_injector(
+        &self,
+        advice_injector: AdviceInjector,
+        host: &mut impl Host,
+    ) -> Result<(), ExecutionError> {
```

Comment:
Let's rename this function to `handle_system_event` and `AdviceInjector` into `SystemEvent` or something similar.